Zend/zend.c

```diff
@@ -502,6 +503,52 @@ ZEND_API void zend_print_zval_r(zval *expr, int indent) /* {{{ */
 }
 /* }}} */
 
+ZEND_API zend_string *zend_zval_to_readable_string(zval *expr)
+{
+try_again:
+	switch (Z_TYPE_P(expr)) {
+		case IS_UNDEF:
+		case IS_ARRAY:
+		case IS_OBJECT:
+		case IS_RESOURCE:
+			return NULL;
+		case IS_NULL:
+			return zend_string_init("null", strlen("null"), 0);
+		case IS_FALSE:
+			return zend_string_init("false", strlen("false"), 0);
+		case IS_TRUE:
+			return zend_string_init("true", strlen("true"), 0);
+		case IS_LONG:
+			return zend_long_to_str(Z_LVAL_P(expr));
+		case IS_DOUBLE:
+			return zend_strpprintf_unchecked(0, "%.*H", (int) EG(precision), Z_DVAL_P(expr));
+		case IS_STRING: {
+			zend_string *str = Z_STR_P(expr);
+			size_t str_len = ZSTR_LEN(str);
+			smart_str readable_str = {0};
+			smart_str_appendc(&readable_str, '"');
+			if (str_len <= READABLE_ZVAL_MAX_STR_LENGTH) {
+				smart_str_appends(&readable_str, ZSTR_VAL(str));
```

Comment:
What's the reason not using `smart_str_append(&readable_str, str)` instead? The string copy below can be avoided with `smart_str_appendl(&readable_str, str, READABLE_ZVAL_MAX_STR_LENGTH)`

For example, with `match("a\x00b") { "a" => ... }`, the error message would say `of value "a"`, even though the actual value is `"a\x00b"`.

In some places, such as anonymous classes, omitting null bytes is deliberate, but probably not great to do here - I'd agree with the escaping of special chars (e.g. null, vertical tabs, control characters) being useful.

Reply:
> What's the reason not using smart_str_append(&readable_str, str) instead?

Probably nothing, except me not knowing the API very well 🙂

> For example, with match("a\x00b") { "a" => ... }, the error message would say of value "a", even though the actual value is "a\x00b".

That's a good point. Wonder how simple that's gonna be. 😕

Reply:
Error messages will be moved to a different discussion. Thanks for your feedback!